kotlinx-coroutines-core/common/src/flow/DataFlow.kt

```diff
@@ -0,0 +1,335 @@
+ *
+ * ```
+ * val aModel = CounterModel()
+ * val bModel = CounterModel()
+ * val sum = aModel.counter.combineLatest(bModel.counter) { a, b -> a + b }
+ * ```
+ *
+ * Conceptually data flow is similar to
+ * [ConflatedBroadcastChannel][kotlinx.coroutines.channels.ConflatedBroadcastChannel]
+ * but it is simpler, because it does not have to implement all the channel APIs.
+ *
+ * All methods of data flow are **thread-safe** and can be safely invoked from concurrent coroutines without
+ * external synchronization.
+ *
+ * ### Operator fusion
+ *
+ * Application of [flowOn], [conflate], and
+ * [buffer] with [CONFLATED][Channel.CONFLATED] or [RENDEZVOUS][Channel.RENDEZVOUS] capacity
+ * has no effect on the data flow.
+ * 
+ * ### Implementation notes
+ *
+ * Data flow implementation is optimized for memory consumption and allocation-freedom. It uses a lock to ensure
+ * thread-safety, but suspending collector coroutines are resumed outside of this lock to avoid dead-locks when
+ * using unconfined coroutines. Adding new collectors has `O(1)` amortized code, but updating a [value] has `O(N)`
+ * cost, where `N` is the number of collectors.
+ */
+@FlowPreview
+public interface DataFlow<T> : Flow<T> {
+    /**
+     * Returns `true` if this data flow is initialized.
+     *
+     * A data flow is initialized when its [value] was set at least once.
+     * A [closed][isClosed] data flow is considered to be initialized.
+     */
+    public val isInitialized: Boolean
+
+    /**
+     * Returns `true` if this data flow is closed.
+     *
+     * See [close].
+     */
+    public val isClosed: Boolean
+
+    /**
+     * Value of this data flow.
+     *
+     * Setting a value that is [equal][Any.equals] to the previous one does nothing. 
+     *
+     * Getting a value from an [uninitialized][isInitialized] flow produces [IllegalStateException].
+     * Getting a value from or setting a value to a [closed][isClosed] flow produces [IllegalStateException].
+     */
+    public var value: T
+
+    /**
+     * Closes this data fl9w.
+     * This is an idempotent operation -- subsequent invocations of this function have no effect and return `false`.
+     * Conceptually, its sets a value to a special "close token" and all active and future collector from
+     * this [Flow] complete either normally (when [cause] is null) or with the corresponding exception that
+     * was specified as a [cause].
+     */
+    public fun close(cause: Throwable? = null): Boolean
+}
+
+/**
+ * Creates an uninitialized [DataFlow].
+ */
+@Suppress("FunctionName")
+@FlowPreview
+public fun <T> DataFlow(): DataFlow<T> = DataFlowImpl(NONE)
+
+/**
+ * Creates a [DataFlow] with a given initial [value].
+ *
+ * This is a shorthand for `DataFlow().apply { this.value = value }`.
+ */
+@Suppress("FunctionName")
+@FlowPreview
+public fun <T> DataFlow(value: T): DataFlow<T> = DataFlowImpl(value ?: NULL)
+
+@SharedImmutable
+private val NONE = Symbol("NONE")
```

Comment:
Please move implementation to a different file (so users won't see it when exploring our public API)